Fix EZP-25021: Implement missing limitation mappers

glye opened the pull request:
> Implements https://jira.ez.no/browse/EZP-25021
> State: Ready for review

Adds the missing limitation mappers, used when editing role policies.

**TODO**
- [x] Owner
- [x] Group
- [x] ~~Node (will need UDW)~~ Moved to https://github.com/ezsystems/repository-forms/pull/63
- [x] ~~Subtree (will need UDW)~~ Moved to https://github.com/ezsystems/repository-forms/pull/63
- [x] State
- [x] NewState
- [x] ParentOwner
- [x] ParentGroup
- [x] ParentClass
- [x] ParentDepth
- [x] Language
- [x] NewSection

glye commented:
Not tested yet due to https://github.com/ezsystems/repository-forms/pull/57#discussion_r44527363

glye commented:
@lolautruche Work in progress, but please see the questions in the descriptions above.

glye commented:
Added Node mapper which is still work in progress, but functioning now. (I think I'll make a `UDWBasedMapper` base class for Node and Subtree, depending on how different they'll have to be given the different input requirements.)

lolautruche commented:
> Shouldn't all these mappers be optional? ("required" = false)

I don't think so. You always need to configure the limitations, like it was the case before in legacy. If you don't need to configure the limitation, then you probably don't need a limitation.

lolautruche commented on bundle/Resources/config/role.yml:
```diff
@@ -95,9 +105,71 @@ services:
+    ezrepoforms.limitation.form_mapper.parentcontenttype:
+        parent: ezrepoforms.limitation.form_mapper.contenttype
+        class: %ezrepoforms.limitation.form_mapper.parentcontenttype.class%
+        arguments: [@ezpublish.api.service.content_type]
```
This is not needed as you already defined `ezrepoforms.limitation.form_mapper.contenttype` as parent.

lolautruche commented on bundle/Resources/config/role.yml:
```diff
@@ -95,9 +105,71 @@ services:
+    ezrepoforms.limitation.form_mapper.parentowner:
+        parent: ezrepoforms.limitation.form_mapper.owner
+        class: %ezrepoforms.limitation.form_mapper.parentowner.class%
+        arguments: [@translator]
```
Same as above. No need to redefine arguments.

lolautruche commented on bundle/Resources/config/role.yml:
```diff
@@ -95,9 +105,71 @@ services:
+    ezrepoforms.limitation.form_mapper.parentgroup:
+        parent: ezrepoforms.limitation.form_mapper.group
+        class: %ezrepoforms.limitation.form_mapper.parentgroup.class%
+        arguments: [@translator]
```
Same

lolautruche commented on bundle/Resources/views/Limitation/udw_limitation_value.html.twig:
```diff
@@ -0,0 +1,14 @@
+>{{"role.policy.limitation.location.udw_button"|trans({}, "ezrepoforms_role")}}</button>
+<div id="{{form.limitationValues.vars.id}}-selected-location">
+    {% if form.limitationValues.vars.value is not empty %}
+        {{ render( controller( "ez_content:viewLocation", {'locationId': form.limitationValues.vars.value, 'viewType': '_platformui_content_name'} ) ) }}
```
Deprecation warning!

You should now use `ez_content:view`, passing `contentId` instead of `locationId`. `viewType` is still valid.

lolautruche commented on lib/Limitation/Mapper/GroupLimitationMapper.php:
```diff
@@ -0,0 +1,53 @@
+    {
+        return [
+            0 => $this->translator->trans('policy.limitation.group.any', [], 'ezrepoforms_role'),
+            1 => $this->translator->trans('policy.limitation.group.self', [], 'ezrepoforms_role'),
```
Why enforcing the keys? Not needed AFAICS.

glye replied:
@lolautruche Not strictly needed to enforce the keys, but IMHO being verbose here has value for others reading the code - the keys correspond to specific values in the limitationtype code. Might consider if we should use constants, though.

lolautruche commented on lib/Limitation/Mapper/OwnerLimitationMapper.php:
```diff
@@ -0,0 +1,56 @@
+            0 => $this->translator->trans('policy.limitation.owner.any', [], 'ezrepoforms_role'),
+            1 => $this->translator->trans('policy.limitation.owner.self', [], 'ezrepoforms_role'),
+            // 2: "Session" is not supported yet, see OwnerLimitationType
+            //2 => $this->translator->trans('policy.limitation.owner.session', [], 'ezrepoforms_role'),
```
Same as above

lolautruche commented on lib/Limitation/Mapper/OwnerLimitationMapper.php:
```diff
@@ -0,0 +1,56 @@
+
+    public function filterLimitationValues(Limitation $limitation)
+    {
+        // 0: "Any" cannot be selected. We don't store it as a limitation value because it's the same as no limitation
```
Then don't add "any" to the list, like for CT or Section

glye commented:
@lolautruche 
> I don't think so. You always need to configure the limitations, like it was the case before in legacy. If you
> don't need to configure the limitation, then you probably don't need a limitation.

> Then don't add "any" to the list, like for CT or Section

I wasn't being clear, sorry. Taken together, these suggestions lead to a problem: If all the settings are required, and we don't have an "Any" choice for Owner and Group, then any content / read policy MUST have owner: "self" and group: "self" set. Which is a problem :smile: (In the current platformui, at least. With a different UI it might be worked around.)

lolautruche commented:
As far as I saw in legacy code, *Any* was always dumped. For ContentType, if you don't select anything, it will act as "any".

glye commented:
@lolautruche Yes, but that's the thing... if the limitation is not optional, you *can't* not select anything. We can *either* have required parameters, *or* use selects without a "no choice/any" option. Not both :smiley:

lolautruche commented:
This is why `required: false` is mostly used

lolautruche commented:
What I meant by saying that limitation config is not optional was for the mappers, not for the user.

glye commented:
Ok, then it's clear :thumbsup:

glye commented:
The Node mapper is working now. (The Subtree mapper should be much of the same.)

lolautruche commented on lib/Limitation/Mapper/LocationLimitationMapper.php:
```diff
@@ -0,0 +1,47 @@
+
+    public function filterLimitationValues(Limitation $limitation)
+    {
+        $limitation->limitationValues = [$limitation->limitationValues];
```
Errr, isn't there something wrong here? :smile:

glye replied:
Possibly, but it works... The form returns the value directly, but the limitation wants it in an array. So I wrap it here, and unwrap it in `mapLimitationForm()`.

lolautruche commented:
I guess you should now implement an abstract UDW mapper

glye commented:
> abstract UDW mapper

Node and Subtree don't have that much in common now, but ok.

lolautruche commented:
They at least share the same template

glye commented:
Added `NewSection` and `NewState` mappers.
Added abstract UDW mapper for the `Node` and `Subtree` mappers.
Found a bug when leaving the UDW mappers unset, working on that.

glye commented:
Review ping @lolautruche @andrerom @bdunogier et al.

andrerom commented:
+1

lolautruche commented on lib/Limitation/Mapper/ContentTypeLimitationMapper.php:
```diff
@@ -33,4 +33,9 @@ protected function getSelectionChoices()
+
+    protected function getChoiceFieldOptions()
+    {
+        return ['required' => false];
```
If we have to add `['required' => false]` everywhere, maybe we should add this as default in abstract mapper

lolautruche commented on lib/Limitation/Mapper/NewObjectStateLimitationMapper.php:
```diff
@@ -0,0 +1,13 @@
+ */
+namespace EzSystems\RepositoryForms\Limitation\Mapper;
+
+class NewObjectStateLimitationMapper extends ObjectStateLimitationMapper
```
If you dont have anything to do here, maybe the class is not needed. Just refer to `ObjectStateLimitationMapper` in service definition

glye replied:
Cool, then the same goes for `ParentContentTypeLimitationMapper`, `ParentGroupLimitationMapper`, and `ParentOwnerLimitationMapper`. Checking...

lolautruche replied:
Less is more :-)

lolautruche commented on lib/Limitation/Mapper/NewSectionLimitationMapper.php:
```diff
@@ -0,0 +1,13 @@
+ */
+namespace EzSystems\RepositoryForms\Limitation\Mapper;
+
+class NewSectionLimitationMapper extends SectionLimitationMapper
```
Same as NewObjectStateLimitationMapper

lolautruche commented on lib/Limitation/Mapper/LocationLimitationMapper.php:
```diff
@@ -0,0 +1,19 @@
+
+class LocationLimitationMapper extends UDWBasedMapper
+{
+    public function filterLimitationValues(Limitation $limitation)
```
I don't see this method used in `USWBasedMapper`. Am I missing something?

lolautruche replied:
Forget it, I'm stupid. I just forgot my own code :blush:

glye replied:
You mean you'd like an empty implementation there, like in `MultipleSelectionBasedMapper`? Sure, can do. Is it CS/policy to implement every part of an interface, even in an abstract class?

glye replied:
Oh, ok :)

lolautruche commented on lib/Limitation/Mapper/OwnerLimitationMapper.php:
```diff
@@ -0,0 +1,39 @@
+        return [
+            1 => $this->translator->trans('policy.limitation.owner.self', [], 'ezrepoforms_role'),
+            // 2: "Session" is not supported yet, see OwnerLimitationType
+            //2 => $this->translator->trans('policy.limitation.owner.session', [], 'ezrepoforms_role'),
```
Please don't commit commented out code. It's fine to have a comment (should be above `return` statement), with what it should be, but commented out code is evil :wink:

lolautruche commented on lib/Limitation/Mapper/ParentContentTypeLimitationMapper.php:
```diff
@@ -0,0 +1,13 @@
+ */
+namespace EzSystems\RepositoryForms\Limitation\Mapper;
+
+class ParentContentTypeLimitationMapper extends ContentTypeLimitationMapper
```
Same as for `NewObjectStateMapper`

lolautruche commented on lib/Limitation/Mapper/ParentDepthLimitationMapper.php:
```diff
@@ -0,0 +1,38 @@
+{
+    /**
+     * @var int The maximum possible depth to use in a limitation
+     * (In legacy the max depth is set in site.ini [RoleSettings] MaxParentDepthLimitation, default: 20)
```
This is not legacy :smiley: . I think you can remove this comment part.

lolautruche commented on lib/Limitation/Mapper/ParentGroupLimitationMapper.php:
```diff
@@ -0,0 +1,13 @@
+ */
+namespace EzSystems\RepositoryForms\Limitation\Mapper;
+
+class ParentGroupLimitationMapper extends GroupLimitationMapper
```
Same as `NewObjectStateMapper`

lolautruche commented on lib/Limitation/Mapper/ParentOwnerLimitationMapper.php:
```diff
@@ -0,0 +1,13 @@
+ */
+namespace EzSystems\RepositoryForms\Limitation\Mapper;
+
+class ParentOwnerLimitationMapper extends OwnerLimitationMapper
```
Same as `NewObjectStateMapper`

glye commented:
Set `['required' => false]` as default in the abstract mapper, and removed it from subclasses
Removed empty classes, and updated service definition accordingly
Removed commented code
Removed legacy comment

lolautruche commented:
+1

bdunogier commented:
So, can we merge this ?

glye commented:
When https://github.com/ezsystems/PlatformUIBundle/pull/433 is approved, yes.

bdunogier commented:
How much does this one depend on ezsystems/platform-ui-bundle#433 ?

glye commented:
Moved the UDW-based mappers (Node and Subtree) to new PR, since https://github.com/ezsystems/PlatformUIBundle/pull/433 will take a while.

bdunogier commented:
And we're in. If not already done, remember to change the issue # in #63, @glye.
